### PR TITLE
Move network code from ./pkg/virt-launcher under ./pkg/network

### PR DESCRIPTION
**Pull request by qinqon**
**What this PR does / why we need it**:
Most network code from ./pkg/virt-launcher is used at virt-handler and virt-launcher to clarify this this PR move all the code th ./pkg/network also helps with the effort of grouping together all the network code.

Top level code movement:
```
Old
pkg/virt-launcher/virtwrap/network/podnic.go
pkg/virt-launcher/virtwrap/network/podnic_test.go
pkg/virt-launcher/virtwrap/network/podinterface_test.go
pkg/virt-launcher/virtwrap/network/network_suite_test.go
pkg/virt-launcher/virtwrap/network/network.go
pkg/virt-launcher/virtwrap/network/network_test.go
pkg/virt-launcher/virtwrap/network/podinterface.go

New
pkg/network/podnic.go
pkg/network/podnic_test.go
pkg/network/podinterface_test.go
pkg/network/network_suite_test.go
pkg/network/network.go
pkg/network/network_test.go
pkg/network/podinterface.go
```

DHCP IPv4 code movement:
```
Old
pkg/virt-launcher/virtwrap/network/dhcp/dhcp_test.go
pkg/virt-launcher/virtwrap/network/dhcp/ethtool.go
pkg/virt-launcher/virtwrap/network/dhcp/dhcp_suite_test.go
pkg/virt-launcher/virtwrap/network/dhcp/socket_listener.go
pkg/virt-launcher/virtwrap/network/dhcp/dhcp.go

New
pkg/network/dhcp/server/server_test.go
pkg/network/link/ethtool.go
pkg/network/dhcp/server/server_suite_test.go
pkg/network/link/socket_listener.go
pkg/network/dhcp/server/server.go

```

DHCP IPv6 code movement:
```
Old
pkg/virt-launcher/virtwrap/network/dhcpv6/dhcpv6_test.go
pkg/virt-launcher/virtwrap/network/dhcpv6/dhcpv6.go
pkg/virt-launcher/virtwrap/network/dhcpv6/conn.go
pkg/virt-launcher/virtwrap/network/dhcpv6/dhcpv6_suite_test.go

New
pkg/network/dhcp/server/serverv6_test.go
pkg/network/dhcp/server/serverv6.go
pkg/network/dhcpv6/conn.go
```

New content extracted from a go file
```
pkg/network/address/reserved_mac.go
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
New packages:
- **./pkg/network/dhcp/server**: Contains the ipv4 and ipv6 DHCP server needed a new package to break the cycle dependency with network handler and DHCP configurators
- **./pkg/network/address**: StaticMasqueradeBridgeMAC is used at podinterface_test to so it creates a cycle dependency, moving it to a package breaks the depdency.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by oshoval — APPROVED** [review]

**Comment by qinqon** [comment]
/cc @EdDev 

This helps to start removing virt-launcher/virt-handler dependencies at network code.

**Comment by oshoval** [comment]
what were the changes ? just rebase right ?

File: pkg/network/link/socket_listener.go [hunk 1]
```diff
@@ -16,7 +16,7 @@
  * Copyright 2020 Red Hat, Inc.
  *
  */
-package dhcp
+package link
```
Comment (EdDev) [inline pkg/network/link/socket_listener.go:19]: Isn't this an UDP socket? If so, how is it fitting a link?
It is also dhcp specific, isn't it? Why move it out?

File: pkg/network/dhcp/server/server.go [hunk 2]
```diff
@@ -17,7 +17,7 @@
  *
  */
 
-package dhcp
+package server
```
Comment (EdDev) [inline pkg/network/dhcp/server/server.go:20]: Why there is a need for a separated package? Could you please explain?

File: pkg/network/dhcp/server/conn.go [hunk 3]
```diff
@@ -16,7 +16,7 @@
  * Copyright 2020 Red Hat, Inc.
  *
  */
-package dhcpv6
+package server
```
Comment (EdDev) [inline pkg/network/dhcp/server/conn.go:19]: There is code here which is DHCPv6 specific but the names do not represent it.
If there is a need to separate dhcp and dhcpv6, this is a good reason for it (context).

File: pkg/network/BUILD.bazel [hunk 4]
```diff
@@ -1,8 +1,55 @@
...
     importpath = "kubevirt.io/kubevirt/pkg/network",
     visibility = ["//visibility:public"],
+    deps = [
+        "//pkg/network/cache:go_default_library",
+        "//pkg/network/dhcp:go_default_library",
+        "//pkg/network/driver:go_default_library",
+        "//pkg/network/errors:go_default_library",
+        "//pkg/network/infraconfigurators:go_default_library",
+        "//pkg/network/link:go_default_library",
+        "//pkg/virt-launcher/virtwrap/api:go_default_library",
```
Comment (EdDev) [inline pkg/network/BUILD.bazel:19]: We should not depend inwards on virt-* packages.

**Review by EdDev — COMMENTED** [review]
I link this path to centralize all network stuff.

Can you please layout the tree structure you moved under `pkg/network` in the description?
Lets observe it at that tree & name level to examine how it looks.

**Comment by EdDev** [comment]
As a following PR or commit, please add license headers to all you touch here.

File: pkg/network/link/socket_listener.go (hunk 1 above)
Comment (qinqon) [reply pkg/network/link/socket_listener.go:19]: Moving back to ./pkg/network/dhcp/server, I don't know why I did move it to link.

File: pkg/network/dhcp/server/server.go (hunk 2 above)
Comment (qinqon) [reply pkg/network/dhcp/server/server.go:20]: First approach was copy it under ./pkg/network/dhcp but problem is that this package depends on the ./pkg/network/driver and that package depend on the dhcp server (the driver package containe the NetworkHandler interface with the StartDHCPServer function) so it creates a cycle, moving it to it's own package fix the issue.

File: pkg/network/dhcp/server/conn.go (hunk 3 above)
Comment (qinqon) [reply pkg/network/dhcp/server/conn.go:19]: What do you think about moving the IPv6 server to serverv6 ?

File: pkg/network/BUILD.bazel (hunk 4 above)
Comment (qinqon) [reply pkg/network/BUILD.bazel:19]: well is where the libvirt Domain structs are living, we cannot remove this dep, maybe problem is that the api.Domain and friends cannot live there, since it's also used by virt-handler.

**Comment by qinqon** [comment]
[push](https://github.com/kubevirt/kubevirt/compare/8e44166b288e37081601028527dad4d98de0591d..d4552437a22a51c0e60bf396d1c8d2029d2d4fb9): Moved socket_listener to pkg/network/dhcp/server

File: pkg/network/dhcp/server/server.go (hunk 2 above)
Comment (EdDev) [reply pkg/network/dhcp/server/server.go:20]: I see. It seems an artificial fix to a problem which should solved from the roots.
But this is a good step forward.

We should break down the driver package.

**Review by EdDev — COMMENTED** [review]
We can categorize kubevirt network code into two: Code that is specific to kubevirt flows (e.g. the phases) and code that is network generic (creating a bridge, reading from the kernel, etc).
I think that this needs to be represented in the `network` package structure.
Specifically, having something like this:
- network:
  - dhcp (for the dhcp servers)
  - link (for anything that represents l2, from devices, through to mac addresses, ethtool, etc...)
  - ip (L3 related helpers, from setting or reading IP addresses, routes, NAT, etc...)
  - setup (for kubevirt setup flows, which probably use all the other sub-packages).

There may also appear packages that abstract other generic functionalities, like: netlink, nftables, iptables. All these should appear also under `network`.

The top level code that was moved here, includes kubevirt flows and therefore we can categorize it as such (although probably many parts may be extracted out to network generics). So I would prefer to see them under a `setup` package or some other name that makes it clear to what category it belongs.

Regarding the `dhcp` package, it makes sense to split the v4 from the v6, and if you suggested in the inline comment to have two sub-packages, one server4 and one server6, it sounds good to me.

File: pkg/network/dhcp/server/server.go (hunk 2 above)
Comment (qinqon) [reply pkg/network/dhcp/server/server.go:20]: It's ok to do that as a follow up ?

**Comment by qinqon** [comment]
[force-push](https://github.com/kubevirt/kubevirt/compare/d4552437a22a51c0e60bf396d1c8d2029d2d4fb9..3b3f8656144410c3e33afeb5499bc807fc05800c): 
- Move DHCP IPv6 server to pkg/network/dhcp/serverv6
- Move phases code to pkg/network/setup
- Move stuff from pkg/network/address to pkg/network/link

File: pkg/network/dhcp/server/server.go (hunk 2 above)
Comment (EdDev) [reply pkg/network/dhcp/server/server.go:20]: Yes, I didn’t meant to fix it now.

File: pkg/network/dhcp/serverv6/values.go [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package serverv6
+
+import "time"
+
+const (
+	infiniteLease = 999 * 24 * time.Hour
```
Comment (EdDev) [inline pkg/network/dhcp/serverv6/values.go:6]: This is strange... Why it is not defined near the usage or the most appropriate source file? It is strange to have a generic file named "values".

File: pkg/network/dhcp/server/values.go [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package server
+
+import "time"
+
+const (
+	infiniteLease = 999 * 24 * time.Hour
```
Comment (EdDev) [inline pkg/network/dhcp/server/values.go:6]: same as the other "values" file.

**Review by EdDev — COMMENTED** [review]
Looks like the last two commits can be squashed or better organized.
A move there is hidden about the mac-reseved addresses, although it is a nit. I would be ok with a simple squash, but if you want it well organized then:
- Move that mac stuff to the link.
- Move the virt code directly to the setup package.
If there are things that live now under /network directly, you can make that another commit to show clearly that it happened.

But again, this is optional for me, up to you.

**Comment by qinqon** [comment]
Addressed @EdDev comments:
- [force-push](https://github.com/kubevirt/kubevirt/compare/3b3f8656144410c3e33afeb5499bc807fc05800c..fd46f61eb9bca135a30e28ca89a7bcd4c4070245): removed values.go  and squashed commits
- [force-push](https://github.com/kubevirt/kubevirt/compare/9aa8f3aca8733d2440f59c7355da48b7712a3a18..0d527544d6e718385bd881c6798751ad1bb1afce): rebase

**Comment by qinqon** [comment]
There is a repeated unit tests name

File: hack/bazel-test.sh [hunk 7]
```diff
@@ -27,4 +29,4 @@ fi
 bazel test \
     --config=${ARCHITECTURE} \
     --features race \
-    --test_output=errors -- //staging/src/kubevirt.io/client-go/... //pkg/... //cmd/... //tests/framework/...
+    --test_output=errors -- ${what}
```
Comment (EdDev) [inline hack/bazel-test.sh:32]: How is this change related?

File: pkg/network/dhcp/server/BUILD.bazel [hunk 8]
```diff
@@ -1,12 +1,37 @@
-load("@io_bazel_rules_go//go:def.bzl", "go_library")
+load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
 
 go_library(
     name = "go_default_library",
-    srcs = ["socket_listener.go"],
+    srcs = [
+        "server.go",
+        "socket_listener.go",
+        "values.go",
```
Comment (EdDev) [inline pkg/network/dhcp/server/BUILD.bazel]: This leaked in between the commits... I saw it in other commits as well.

File: pkg/network/dhcp/serverv6/values.go (hunk 5 above)
Comment (EdDev) [reply pkg/network/dhcp/serverv6/values.go]: I still see this fine in commit `net, dhcp: Move IPv6 server to serverv6 package`

File: hack/bazel-test.sh (hunk 7 above)
Comment (qinqon) [reply hack/bazel-test.sh]: Well is not :-/

File: hack/bazel-test.sh [hunk 9]
```diff
@@ -27,4 +27,4 @@ fi
 bazel test \
     --config=${ARCHITECTURE} \
     --features race \
-    --test_output=errors -- //staging/src/kubevirt.io/client-go/... //pkg/... //cmd/... //tests/framework/...
+    --test_output=errors -- //staging/src/kubevirt.io/client-go/... //pkg/... //cmd/... //tests/framework/...
```
Comment (EdDev) [inline hack/bazel-test.sh]: This is added here and removed in the next commit. Leaving it up to you if you want to clean it up properly.

**Review by EdDev — APPROVED** [review]
Thank you!

The inline nit is left up to you.

File: hack/bazel-test.sh (hunk 9 above)
Comment (qinqon) [reply hack/bazel-test.sh]: If there is something asked from maintainer I will include this fix.

**Comment by qinqon** [comment]
/retest

TLS hiccup
```
 net/http: TLS handshake timeout.
```
